MultiQuestensions/src/main.cpp

```diff
@@ -47,60 +52,68 @@ static void HandlePreviewBeatmapPacket(MultiplayerExtensions::Beatmaps::PreviewB
 }
 
 
 
 // LobbyPlayersDataModel Activate
-MAKE_HOOK_OFFSETLESS(LobbyPlayersActivate, void, GlobalNamespace::LobbyPlayersDataModel* self) {
+MAKE_HOOK_MATCH(LobbyPlayersActivate, &LobbyPlayersDataModel::Activate, void, LobbyPlayersDataModel* self) {
     lobbyPlayersDataModel = self;
     LobbyPlayersActivate(lobbyPlayersDataModel);
 }
 
 // LobbyPlayersDataModel SetLocalPlayerBeatmapLevel
-MAKE_HOOK_OFFSETLESS(LobbyPlayersSetLocalBeatmap, void, GlobalNamespace::LobbyPlayersDataModel* self, Il2CppString* levelId, GlobalNamespace::BeatmapDifficulty beatmapDifficulty, GlobalNamespace::BeatmapCharacteristicSO* characteristic) {
-    GlobalNamespace::IPreviewBeatmapLevel* localPreview = self->beatmapLevelsModel->GetLevelPreviewForLevelId(levelId);
+MAKE_HOOK_MATCH(LobbyPlayersSetLocalBeatmap, &LobbyPlayersDataModel::SetLocalPlayerBeatmapLevel, void, LobbyPlayersDataModel* self, Il2CppString* levelId, BeatmapDifficulty beatmapDifficulty, BeatmapCharacteristicSO* characteristic) {
+    IPreviewBeatmapLevel* localPreview = self->beatmapLevelsModel->GetLevelPreviewForLevelId(levelId);
     if (localPreview != nullptr) {
-        getLogger().info("Local user selected song '", levelId, "'.");
+        getLogger().info("Local user selected song '%s'.", to_utf8(csstrtostr(levelId)).data());
         MultiQuestensions::Beatmaps::PreviewBeatmapStub* preview = CRASH_UNLESS(il2cpp_utils::New<MultiQuestensions::Beatmaps::PreviewBeatmapStub*>(localPreview));
 
         if (preview->levelHash != nullptr) {
             if (self->get_localUserId() == self->get_hostUserId()) {
-                sessionManager->SetLocalPlayerState(il2cpp_utils::createcsstr(beatmapDownloadedState), true);
+                sessionManager->SetLocalPlayerState(il2cpp_utils::newcsstr(beatmapDownloadedState), true);
             }
-
-            packetManager->Send(reinterpret_cast<LiteNetLib::Utils::INetSerializable*>(preview->GetPacket(characteristic->get_serializedName(), beatmapDifficulty)));
-            self->menuRpcManager->SelectBeatmap(GlobalNamespace::BeatmapIdentifierNetSerializable::New_ctor(levelId, characteristic->get_serializedName(), beatmapDifficulty));
-            self->SetPlayerBeatmapLevel(self->get_localUserId(), reinterpret_cast<GlobalNamespace::IPreviewBeatmapLevel*>(preview), beatmapDifficulty, characteristic);
+            MultiplayerExtensions::Beatmaps::PreviewBeatmapPacket* Packet = preview->GetPacket(characteristic->get_serializedName(), beatmapDifficulty);
+            packetManager->Send(reinterpret_cast<LiteNetLib::Utils::INetSerializable*>(Packet));
+            self->menuRpcManager->SelectBeatmap(BeatmapIdentifierNetSerializable::New_ctor(levelId, characteristic->get_serializedName(), beatmapDifficulty));
+            if (il2cpp_functions::class_is_assignable_from(classof(IPreviewBeatmapLevel*), classof(MultiQuestensions::Beatmaps::PreviewBeatmapStub*))) {
```

Comment:
Consider using il2cpp_utils::IsAssignableFrom instead, should be friendlier to use

Reply:
Or even one of the cast helper functions, assuming you are casting